packages/local_auth/local_auth_android/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java

```diff
@@ -251,39 +250,42 @@ private void stopAuthentication(Result result) {
   /*
    * Returns biometric types available on device
    */
-  private void getAvailableBiometrics(final Result result) {
+  private void getEnrolledBiometrics(final Result result) {
     try {
       if (activity == null || activity.isFinishing()) {
         result.error("no_activity", "local_auth plugin requires a foreground activity", null);
         return;
       }
-      ArrayList<String> biometrics = getAvailableBiometrics();
+      ArrayList<String> biometrics = getEnrolledBiometrics();
       result.success(biometrics);
     } catch (Exception e) {
       result.error("no_biometrics_available", e.getMessage(), null);
     }
   }
 
-  private ArrayList<String> getAvailableBiometrics() {
+  private ArrayList<String> getEnrolledBiometrics() {
     ArrayList<String> biometrics = new ArrayList<>();
     if (activity == null || activity.isFinishing()) {
       return biometrics;
     }
-    PackageManager packageManager = activity.getPackageManager();
-    if (Build.VERSION.SDK_INT >= 23) {
-      if (packageManager.hasSystemFeature(PackageManager.FEATURE_FINGERPRINT)) {
-        biometrics.add("fingerprint");
-      }
+    // If no hardware is present, an empty list is returned.
+    if (!hasBiometricHardware()) {
+      return biometrics;
     }
-    if (Build.VERSION.SDK_INT >= 29) {
-      if (packageManager.hasSystemFeature(PackageManager.FEATURE_FACE)) {
-        biometrics.add("face");
-      }
-      if (packageManager.hasSystemFeature(PackageManager.FEATURE_IRIS)) {
-        biometrics.add("iris");
-      }
+    // If no biometrics are enrolled, only "undefined" is returned.
```

Comment:
Since this is a bizarre implementation anyway, and you're already adjusting the method channel slightly, let's just split out a separate method channel call that's backed by `canAuthenticateWithBiometrics`. Then this method can only return enrolled biometrics, eliminating "undefined", and the Dart code that currently relies on this special value can instead use that new method channel call directly.

Reply:
> In the current implementation, the `undefined` value gets ignored on the dart side, both for Android and iOS. As such, I don't believe there is any dart logic that is currently depending on this value, it's just a weird detail being sent back by the native iOS implementation.

It's not ignored, it's how [this](https://github.com/flutter/plugins/blob/37c97713ed6cb1922a76e94cb846760d5e3e2d7a/packages/local_auth/local_auth_android/lib/local_auth_android.dart#L51-L53) works.

> Should we just consider removing `undefined` alltogether in both implementations instead?

Yes, I'm suggesting that we remove "undefined" in favor of a new platform channel method that allows direct querying of whether the device supports biometrics. I.e., have a `deviceSupportsBiometrics` method call, and replace the code I linked to above with a call to that.

(I'd like to do it for iOS at some point, but for this PR we should just do Android).

> Additionally, we could add a second method to the interface (e.g. `getPresentBiometrics`) that just returns a list of all present biometric types, disregarding wether they are enrolled or not, so that this functionality remains available on Android. (But this would then need to also be implemented on iOS)

I'm not sure what you mean by "this functionality"; the functionality that we need to preserve is just `deviceSupportsBiometrics` working.

Reply:
> It's not ignored, it's how [this](https://github.com/flutter/plugins/blob/37c97713ed6cb1922a76e94cb846760d5e3e2d7a/packages/local_auth/local_auth_android/lib/local_auth_android.dart#L51-L53) works.

Am I missing something? [here](https://github.com/flutter/plugins/blob/37c97713ed6cb1922a76e94cb846760d5e3e2d7a/packages/local_auth/local_auth_android/lib/local_auth_android.dart#L70) the value seems to be ignored (same on iOS), which results in an empty list regardless of whether the `undefined` value was returned by the native side, which is what `deviceSupportsBiometrics` uses there.

> the functionality that we need to preserve is just deviceSupportsBiometrics working.

Right, which needs to return `true` even if those biometrics are not enrolled, which would then mean `deviceSupportsBiometrics` is broken on iOS as well, as `undefined` is ignored there too. 

conclusion being that `deviceSupportsBiometrics` needs its own method channel call on both platforms, correct?

Reply:
> Am I missing something? [here](https://github.com/flutter/plugins/blob/37c97713ed6cb1922a76e94cb846760d5e3e2d7a/packages/local_auth/local_auth_android/lib/local_auth_android.dart#L70) the value seems to be ignored (same on iOS)

🤦🏻 

Even after all the discussion about this when you initially extracted the default method channel implementation, I still missed that in review. That's a regression from the refactor: `deviceSupportsBiometrics` is not supposed to be calling the Dart function, it's supposed to be directly invoking the method channel call:

https://github.com/flutter/plugins/commit/dd2cd964a136b2191ec7887004ac0ddf0657ee89#diff-41eeb8757de140d2fbb1c3650b056c1721926ef417103bf2084a3cc48d1f504fR126-R129

> Right, which needs to return true even if those biometrics are not enrolled, which would then mean deviceSupportsBiometrics is broken on iOS as well

Yes, the regression broke iOS too. And also any third-party implementations that relied on this behavior (likely that's none, but it's still wrong). We need to fix this in all three packages.

For the default method channel we need to restore calling the method channel directly. For iOS and Android, we should add a new method. If we didn't already know how confusing the `undefined` implementation was, the fact that I was specifically watching for it in the refactor and *still* missed that the final version wasn't right certainly confirms it.

Reply:
Alright, can do. I'll fix this for Android now, and I'll try submit a fix for the iOS implementation tomorrow.

Reply:
iOS fix submitted in 
- #5329